File: 백준/Silver/1991. 트리 순회/트리 순회.py
import sys


class treeNode:
    def __init__(self,data,left,right):
        self.data=data
        self.left=left
        self.right=right


def inorder(root):
    if root.left != '.':
        inorder(tree[root.left])

    print(root.data, end="")

    if root.right != '.':
        inorder(tree[root.right])

def preorder(root):
    print(root.data, end="")

    if root.left != '.':
        preorder(tree[root.left])
    if root.right!= '.':
        preorder(tree[root.right])

def postorder(root):
    if root.left !='.':
        postorder(tree[root.left])

    if root.right!='.':
        postorder(tree[root.right])

    print(root.data, end="")

n=int(input())
tree={}


for _ in range(n):
    x,y,z=map(str,sys.stdin.readline().split())
    tree[x]=treeNode(data=x,left=y,right=z)

preorder(tree['A'])
print()
inorder(tree['A'])
print()
postorder(tree['A'])
print()

File: 백준/Silver/1991. 트리 순회/트리 순회.cc
#include <iostream>
using namespace std;

struct Node {
    char data;
    Node *left;
    Node *right;
};

int N;
Node arr[30];

void prefix_(Node *start) {
    
    if(start!=nullptr) {
        cout << start->data;
        prefix_(start->left);
        prefix_(start->right);
    }
}

void inorder_(Node *start) {
    if(start!=nullptr) {
        inorder_(start->left);
        cout << start->data;
        inorder_(start->right);
    }
}

void postorder_(Node *start) {
    if(start!=nullptr) {
        postorder_(start->left);
        postorder_(start->right);
        cout << start->data;
    }
}
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    
    cin >> N;
    for(int i=0;i<N;i++) {
        char a,b,c;
        cin >> a >> b >> c;
        arr[a-'A'].data=a;
        
        if(b!='.') {
            arr[a-'A'].left=&arr[b-'A'];
        }
        
        if(c!='.') {
            arr[a-'A'].right=&arr[c-'A'];
        }
    }
    
    prefix_(&arr[0]);
    cout << "\n";
    inorder_(&arr[0]);
    cout << "\n";
    postorder_(&arr[0]);
    cout << "\n";
    
    return 0;
}
